### DISCUSSION TITLE
Are there tradeoffs with `AllowDirect: true`?

### DISCUSSION THREAD

**Discussion by jgaskins** [nats-io/nats-server#3984] [Q&A]
Is there ever a reason to avoid `AllowDirect: true` when creating a stream? Are there tradeoffs being made?

Just trying to understand why it's optional and, from what I can tell, not the default (other than for KV).

**Comment by derekcollison** [comment]
Used for internal usage only atm.

**Reply by jgaskins** [reply]
I maintain [a NATS client](https://github.com/jgaskins/nats), so the internal usage is what I'm actually trying to understand. 😄 I looked at [the ADR for the `DIRECT.GET` API](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-31.md) and I see that the server sets it based on whether `max_msgs_per_subject` is set (presumably the reason this became a required setting for KV to distribute reads), but I haven't been able to uncover why it isn't set for all streams. I assume there are tradeoffs involved.

**Reply by derekcollison** [reply]
My answer is wrong, I thought you were talking about direct consumers which is a whole different thing.

Yes you can edit a stream to enable.

**Comment by kolypto** [comment]
I [grepped the code for `AllowDirect`](https://github.com/search?q=repo%3Anats-io%2Fnats-server%20AllowDirect&type=code) and found that it influences the way clusters work: probably allocating more resources because of unoptimized access?

I'd assume that keeping it `AllowDirect: false` is easier on server resources
